Dockerfile

```diff
@@ -0,0 +1,29 @@
+FROM node:20-alpine as builder
+
+# Install vendor dependencies
+WORKDIR /source/vendor
+COPY vendor/package.json package.json
+COPY vendor/package-lock.json package-lock.json
+RUN npm ci
+
+# Install app dependencies
+WORKDIR /source
+COPY package.json package.json
+COPY package-lock.json package-lock.json
+RUN npm ci
+
+# Copy project files and run necessary symlinks npm needs
+COPY app app
+COPY vendor vendor
+RUN npm ci
+
+# Buld the app
+COPY tsconfig.json tsconfig.json
+RUN npm run build
```

Comment:
A few changes in here:
1. Path basing can be confusing, so I recommend being quite explicit.  `./` for "root of Docker build context" and `/app/` for the target directory in the container as an absolute path.  You could do `/source/` for the target directory in the container (or anything else really), but `/app` I think is a fairly common choice for "where in the container do you put your app" so people may look there first.  Another common one is `workspace` I think.
2. I removed the excess `npm ci` because it shouldn't be necessary.  Let me know if it is needed for some reason.
3. Add running of the vendor script, which appears to be missing.
4. Be explicit about which folders are copied.  This is important because if you just copy the whole `./app` folder then you'll *also* copy over the `js` files.  Our build script deletes them, but one can imagine other folders with generated data in them later, and being explicit ensures we don't accidentally copy anything we shouldn't be.

```suggestion
# Install vendor dependencies
COPY ./vendor/package.json /app/vendor/package.json
COPY ./vendor/package-lock.json /app/vendor/package-lock.json
WORKDIR /app/vendor
RUN npm ci

# Install app dependencies
COPY ./package.json /app/package.json
COPY ./package-lock.json /app/package-lock.json
WORKDIR /app
RUN npm ci

# run vendor script
COPY ./vendor/ /app/vendor/
WORKDIR /app/build
RUN npm run vendor

# Buld the app
COPY ./tsconfig.json /app/tsconfig.json
COPY ./index.html /app/app/index.html
COPY ./app/css/ /app/app/css/
COPY ./app/img/ /app/app/img/
COPY ./app/ts/ /app/app/ts/
WORKDIR /app
RUN npm run build
```